### ISSUE TITLE
[MQTT-eventbus] Sonos Events Published as "Undefined"

### ISSUE THREAD

**Issue by AcidGambit** [openhab/openhab#4902]
I'm currently using the snapshot version of OH2 with the MQTT binding. I've setup MQTT with Event Bus configuration. The issue is that I am seeing real values for a Sonos control item in my events.log, but "Undefined" is published to MQTT.

For example, in my events.log, I see the following when I pause a player:

2016-12-04 22:28:21.388 [ItemStateChangedEvent ] - sonosJOControl changed from PLAY to PAUSE

But the following is what I see on the related topic for the same occurrence (using mosquitto_sub):

XXXXXX/out/sonosJOControl/state Undefined

I would expect to see "PAUSE" published to the topic.

**Comment by 9037568** [comment]
Can you post your openhab.log and events.log files?

**Comment by AcidGambit** [comment]
Yes, they're attached, though they don't look particularly interesting (do you need debug?):

[logs.zip](https://github.com/openhab/openhab/files/657221/logs.zip)

As for the Control items, I see entries like this when I subscribe to the relevant topics in MQTT:

out/sonosJOControl/state Undefined

**Comment by martinvw** [comment]
Can you elaborate a little bit more on the "MQTT with Event Bus configuration", I am using OH1 with the mqtt persistence to sync some data to my OH2 instance.

But is this the same you are using or did you set it up differently?

So what binding are you using and what version of the binding 1.9 or 2.0?

**Comment by AcidGambit** [comment]
Yes, I am using OH2 to publish to a OH1 instance.  This is my first time using MQTT event bus because I am going to a new architecture with two OH instances on different networks.

In the OH2 instance (I am running the snapshot version and updated a few days ago), I only have a single MQTT extension choice available in Paper UI:

binding-mqtt1 - 1.9.0.SNAPSHOT

I should also add that only the Sonos Player items (e.g., sonosJOControl) are publishing UNDEFINED values.   All other items are published properly.

**Comment by martinvw** [comment]
Sorry some final details which I am interested in:

- its about the 1.9 binding in the category persistence?
- could you enable debug logging for the package org.openhab.io.transport.mqtt and corn package org.openhab.persistence.mqtt
- Could you give me the item or thing definition for the item you try to persist or has it been crated via the paper ui? I am interested in the type of sonosJOControl, because I see no clear reason why it fails
- do you have the problem in another openHAB instance of also when subscribing directly via command line to the mqtt topic?

**Comment by 9037568** [comment]
Yes, debug logs are always best.

**Comment by AcidGambit** [comment]
I'll answer each of your questions below:

1.  MQTT persistence binding version in Paper UI shows as:  persistence-mqtt - 1.9.0.SNAPSHOT.  It wasn't previously enabled, but I have enabled it now.

2.  I have enabled debug logging for both and the logs are attached, though I deleted personal information (e.g., IP addresses).

3.  I used Paper UI to create the items, but the "Control" items are "Player" type items with, for example, the following channel:  "sonos:CONNECTAMP:RINCON_XXXXX:control"  The other items are Number and String type.  The linking information in the beginning of events.log also identifies how each Item is defined.

4.  I haven't bothered to setup the other openHAB instance receiving the Sonos items because I debug the setup using mosquitto_sub before I setup openHAB.  That said, I have my other openHAB instance subscribed to a String type item from another addon (Harmony Hub).



[Archive.zip](https://github.com/openhab/openhab/files/659040/Archive.zip)

**Comment by AcidGambit** [comment]
From my end, the logs didn't look particularly interesting.  Let me know if TRACE level logging will help.

**Comment by martinvw** [comment]
Sorry the problem is clear to me, you are not using the exact same way as I do.

You are using the MQTT binding with the option _Event Bus Binding Configuration_ like you mentioned in the title and your first post. https://github.com/openhab/openhab/wiki/MQTT-Binding#event-bus-binding-configuration

And I use the MQTT Persistence binding: https://github.com/openhab/openhab/wiki/MQTT-Persistence

The player type which is used by the SONOS binding is a OH2 specific type and MQTT  1.9 binding you use is not aware of the new Player type, so it does not surprise me that this does not work.

Looking at the code this might even be caused by the compatibility layer, but what can the compatibility layer do about mapping a new OH2 type to a not existing one in the OH1-world.

You could try your setup also with the MQTT Persistence binding but it might well suffer from the same problems.

A viable workaround could be storing the player state also in StringItem via a rule and posting this string item with the MQTT eventbus.

**Comment by AcidGambit** [comment]
Thank you, interestingly, the same issue occurs if I change the item definition to String.  But, per your suggestion, adding a rule that changes an independent string item works.

**Comment by 9037568** [comment]
I see in the log:
>2016-12-17 14:25:15.146 [ERROR] [core.karaf.internal.FeatureInstaller] - Failed installing 'openhab-binding-mqtt': No matching features for openhab-binding-mqtt/0.0.0

@watou could this be related to the recent feature renaming that was done?

**Comment by watou** [comment]
> could this be related to the recent feature renaming that was done?

I'm fairly sure that's it.  Make sure to put a `1` at the end of 1.x binding feature names.  So you might have
```
bindings = mqtt1
```
In addons.cfg, for example. This applies to recent nightly builds.

**Comment by 9037568** [comment]
Please let us know how watou's suggested fix works, @AcidGambit .

**Comment by AcidGambit** [comment]
Sorry for not responding - I can't check until the weekend.

**Comment by AcidGambit** [comment]
My addons.cfg does not have mqtt identified.  Instead, "binding" is still commented out.  Is this because I installed MQTT through Paper UI?

**Comment by 9037568** [comment]
Not sure on that question. Possibly asking in the user community would yield an answer.

**Comment by AcidGambit** [comment]
I apologize for not getting back to you.  Currently running the RC snapshot and the same issue is there, even with defining the addon as mqtt1.  That said, the workaround of converting the player state to a string works.

**Comment by nanosonde** [comment]
Me too.

I want to connect an OH2 instance to another smart homesystem using MQTT and listening on OH2 event bus by configuring mqtt-eventbus stuff.
However, for the SONOS binding I only get something like this:
/openHAB/out/SonosPLAYBAR_Control/state Undefined

OH2 event log says:
`
2017-05-11 16:12:16.565 [ItemCommandEvent          ] - Item 'SonosPLAYBAR_Control' received command PAUSE
2017-05-11 16:12:16.569 [ItemStateChangedEvent     ] - SonosPLAYBAR_Control changed from PLAY to PAUSE
`

**Comment by nanosonde** [comment]
Hi!
I have started to debug OH2 with the SONOS binding 2.x and the MQTT binding 1.x.

The undefined message is caused by the SONOS binding.
It may be related to recent SONOS FW updates. I am not sure.
There is a now a new state called "TRANSITIONING" while changed from PLAYING to STOPPED or vice versa. As you can see in the code snippet from ZonePlayerHandler.java, this state is handled in the if-else() tree. This means that the variable newState keeps the value UnDefType.UNDEF.
After the TRANSITIONING state, the final state PLAYING or PAUSED is immediately fired afterwards.
However, for whatever reason, the binding does not emit the second "real" event. 
I assume that there was no TRANSITIONING state in the past and so code is expecting to only have one event with either PLAYING or STOPPED and not two events following each other. 
The event "TRANSITIONING" is causing the UNDEFINED message on the event bus.

Unfortunately, I do not know how to fix it.

`
        State newState = UnDefType.UNDEF;
        switch (channeldD) {
            case STATE:
                if (stateMap.get("TransportState") != null) {
                    newState = new StringType(stateMap.get("TransportState"));
                }
                break;
            case CONTROL:
                if (stateMap.get("TransportState") != null) {
                    if (stateMap.get("TransportState").equals("PLAYING")) {
                        newState = PlayPauseType.PLAY;
                    } else if (stateMap.get("TransportState").equals("STOPPED")) {
                        newState = PlayPauseType.PAUSE;
                    } else if (stateMap.get("TransportState").equals("PAUSED_PLAYBACK")) {
                        newState = PlayPauseType.PAUSE;
                    }
                }
                break;
`

**Comment by 9037568** [comment]
Good analysis. It also appears that there is no [PlayPauseType ](https://github.com/eclipse/smarthome/blob/0c3fce4d3df5925610f45a59f6c28911823fceff/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/PlayPauseType.java#L21) that would correspond to "TRANSITIONING". There is only PLAY or PAUSE.

But it appears that any fix will need to take place in [ESH](http://github.com/eclipse/smarthome/issues).

**Comment by nanosonde** [comment]
So what do you suggest? Opening a new issue in ESH?

**Comment by 9037568** [comment]
Yes, that's what I'd suggest.

**Comment by nanosonde** [comment]
I have opened a new [issue](https://github.com/eclipse/smarthome/issues/3491) in ESH.
